services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureRequestObjectExecutor.java

```diff
@@ -0,0 +1,508 @@
+            }
+        }
+
+        // check whether request object can be parsed successfully
+        JWSInput input;
+        try {
+            input = new JWSInput(retrievedRequestObject);
+            parsedRequestObject = JsonSerialization.readValue(input.getContent(), AuthorizationEndpointRequestObject.class);
+        } catch (JWSInputException | IOException e) {
+            ClientPolicyLogger.log(logger, "failed to parse request object.");
+            throw new ClientPolicyException(OAuthErrorException.INVALID_REQUEST, "Invalid request object");
+        }
+
+        if (logger.isTraceEnabled()) {
+            DumpQueryParameters();
+            DumpRequestObject();
+        }
+
+        // check whether scope exists in both query parameter and request object
+        if (params.getFirst(OIDCLoginProtocol.SCOPE_PARAM) == null || parsedRequestObject.getScope() == null) {
+            ClientPolicyLogger.log(logger, "scope does not exists.");
+            throw new ClientPolicyException(OAuthErrorException.INVALID_REQUEST, "Missing parameter : scope");
+        }
+
+        // check whether "exp" claim exists
+        if (parsedRequestObject.getExp() == null) {
+            ClientPolicyLogger.log(logger, "exp claim not incuded.");
+            throw new ClientPolicyException(INVALID_REQUEST_OBJECT, "Missing parameter : exp");
+        }
+
+        // check whether request object not expired
+        long exp = parsedRequestObject.getExp().longValue();
+        if (Time.currentTime() > exp) { // TODO: Time.currentTime() is int while exp is long...
+            ClientPolicyLogger.log(logger, "request object expired.");
+            throw new ClientPolicyException(INVALID_REQUEST_OBJECT, "Request Expired");
+        }
+
+        // check whether "aud" claim exists
+        String[] aud = parsedRequestObject.getAudience();
+        if (aud == null) {
+            ClientPolicyLogger.log(logger, "aud claim not incuded.");
+            throw new ClientPolicyException(INVALID_REQUEST_OBJECT, "Missing parameter : aud");
+        }
+
+        // check whether "aud" claim points to this keycloak as authz server
+        String iss = Urls.realmIssuer(session.getContext().getUri().getBaseUri(), session.getContext().getRealm().getName());
+        if (!Arrays.asList(aud).contains(iss)) {
+            ClientPolicyLogger.log(logger, "aud not points to the intended realm.");
+            throw new ClientPolicyException(INVALID_REQUEST_OBJECT, "Invalid parameter : aud");
+        }
+
+        // confirm whether all parameters in query string are included in the request object, and have the same values
+        // argument "request" are parameters overridden by parameters in request object
+        if (KNOWN_REQ_PARAMS.stream().filter(s->params.containsKey(s)).anyMatch(s->!isSameParameterIncluded(s, params.getFirst(s), parsedRequestObject))) {
```

Comment:
The algorithm here and in the "isSameParameterIncluded" is a bit uneffective IMO. It iterates through all the OIDC supported parameters and calls very big amount of "if" and "equals" inside "isSameParameterIncluded" . Also will require change in the code upon any new supported OIDC parameter and whole "isSameParameterIncluded" does not scale etc.

Maybe it is possible to leverage the fact that OIDC query parameter names are the same as parameters inside "request" object. Instead of adding all the passed "request" parameters as properties of the AuthorizationEndpointRequestObject, it may be better to just keep lots of those properties as generic properties in the "otherClaims" map. Then the algorithm should be possibly made more effective like for example something like this:
```
for (Map.Entry<String, List<String>> param : params.entrySet()) {
    if (KNOWN_REQ_PARAMS.contains(param.getKey()) {
        String requestValue = parsedRequestObject.getOtherClaim().get(param.getKey());
        if (requestValue == null || requestValue.equals(param.getValue() {
            // Throw error here. It can contain the name of the failing parameter in the trace logging message, which is advantage IMO
        }
    }
}
```

Reply:
@mposolda Just mentioned above, Instead of using dedicated AuthorizationEndpointRequestObject, I'll use bearer JsonNode for parsedRequestObject, which can simplify `isSameParameterIncluded` as you pointed out.